x509/revocation/google/google.go

```diff
@@ -0,0 +1,381 @@
+package google
+
+import (
+	"archive/zip"
+	"bytes"
+	"crypto"
+	"crypto/rsa"
+	"crypto/sha1"
+	"crypto/sha256"
+	"encoding/binary"
+	"encoding/hex"
+	"encoding/json"
+	"encoding/xml"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math/big"
+	"net/http"
+	"net/url"
+
+	"github.com/zmap/zcrypto/x509"
+)
+
+// CRLSet - data structure for storing CRLSet data, used by methods below
+type CRLSet struct {
+	Version      string
+	IssuerLists  map[string]*IssuerList
+	Sequence     int
+	NumParents   int
+	BlockedSPKIs []string
+}
+
+// IssuerList - list of revoked certificate entries for a given issuer
+type IssuerList struct {
+	SPKIHash string // SHA256 of Issuer SPKI
+	Entries  []*Entry
+}
+
+// Entry - entry for a single certificate
+type Entry struct {
+	SerialNumber *big.Int
+}
+
+// FetchAndParse - fetch from distribution point, parse to CRLSet struct as defined above
+func FetchAndParse() (*CRLSet, error) {
+	crlSetReader, version, err := fetch()
+	if err != nil {
+		return nil, err
+	}
+	return Parse(crlSetReader, version)
+}
+
+// Check - Given a parsed CRLSet, check if a given cert is present
+func (crlSet *CRLSet) Check(cert *x509.Certificate, issuerSPKIHash string) *Entry {
+	issuersRevokedCerts := crlSet.IssuerLists[issuerSPKIHash]
+	if issuersRevokedCerts == nil { // no entries for this issuer
+		return nil
```

Comment:
Would it be useful to distinguish found-but-not-revoked from no-CRL-for-this-issuer?

Reply:
I can't think of a reason that it would be necessary. Returning at line 58 effectively means that "there are other certs signed by the same parent that are now revoked," but once we run the scanner on all certs it would be straightforward to determine this in other ways. (Line 57 being true doesn't imply that there's no revocation data available, just that there are no certs revoked by CRLSet with this issuer.) Maybe there's some situation I haven't considered though?